evm/src/extension_tower.rs

```diff
@@ -0,0 +1,1198 @@
+        let x0 = U512(self.val.0[..4].try_into().unwrap());
+        let x1 = U512(self.val.0[4..].try_into().unwrap());
+        let y0 = U512(other.val.0[..4].try_into().unwrap());
+        let y1 = U512(other.val.0[4..].try_into().unwrap());
+
+        let z00 = BLS381 {
+            val: x0.saturating_mul(y0) % BLS_BASE,
+        };
+        let z01 = BLS381 {
+            val: x0.saturating_mul(y1),
+        };
+        let z10 = BLS381 {
+            val: x1.saturating_mul(y0),
+        };
+        let z11 = BLS381 {
+            val: x1.saturating_mul(y1),
+        };
+
+        z00 + (z01 + z10).lsh_256() + z11.lsh_512()
+    }
+}
+
+impl FieldExt for BLS381 {
+    const ZERO: Self = BLS381 { val: U512::zero() };
+    const UNIT: Self = BLS381 { val: U512::one() };
+    fn inv(self) -> BLS381 {
+        let exp = BLS_BASE - 2;
+        let mut current = self;
+        let mut product = BLS381 { val: U512::one() };
+
+        for j in 0..512 {
+            if exp.bit(j) {
+                product = product * current;
+            }
+            current = current * current;
+        }
+        product
+    }
+}
+
+#[allow(clippy::suspicious_arithmetic_impl)]
+impl Div for BLS381 {
+    type Output = Self;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        self * rhs.inv()
+    }
+}
+
+/// The degree 2 field extension Fp2 is given by adjoining i, the square root of -1, to BN254
+/// The arithmetic in this extension is standard complex arithmetic
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Fp2<T>
+where
+    T: FieldExt,
+{
+    pub re: T,
+    pub im: T,
+}
+
+impl<T> Distribution<Fp2<T>> for Standard
+where
+    T: FieldExt,
+    Standard: Distribution<T>,
+{
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Fp2<T> {
+        let (re, im) = rng.gen::<(T, T)>();
+        Fp2 { re, im }
+    }
+}
+
+impl<T: FieldExt> Add for Fp2<T> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Fp2 {
+            re: self.re + other.re,
+            im: self.im + other.im,
+        }
+    }
+}
+
+impl<T: FieldExt> Neg for Fp2<T> {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Fp2 {
+            re: -self.re,
+            im: -self.im,
+        }
+    }
+}
+
+impl<T: FieldExt> Sub for Fp2<T> {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        Fp2 {
+            re: self.re - other.re,
+            im: self.im - other.im,
+        }
+    }
+}
+
+impl<T: FieldExt> Mul for Fp2<T> {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        Fp2 {
+            re: self.re * other.re - self.im * other.im,
+            im: self.re * other.im + self.im * other.re,
+        }
+    }
+}
+
+impl<T: FieldExt> Fp2<T> {
+    /// This function scalar multiplies an Fp2 by an BN254
+    pub fn scale(self, x: T) -> Self {
+        Fp2 {
+            re: x * self.re,
+            im: x * self.im,
+        }
+    }
```

Comment:
I suggest just implementing `Mul<Rhs = T> for Fp2<T>` and `Mul<Rhs = Fp2<T>> for T`. That way you an use the normal `*` syntax.